### DISCUSSION TITLE
MeshGL::faceID behavior

### DISCUSSION THREAD

**Discussion by ramcdona** [elalish/manifold#567] [General]
I create several MeshGL's, each with triangles assigned sequential faceID's.  MeshGL A has faceID's [0,99], B has [100,199] etc.  This way, the faceID's can be used as indexes into an array of face properties.

After converting the MeshGL's to manifolds, combining through Boolean operations, and doing a GetMeshGL(), the output ID's no longer seem suitable for this purpose.  It is almost as if each MeshGL's ID's have been re-set to count from zero.

Am I misunderstanding the intended function of faceID?  Is there a bug?

**Comment by elalish** [comment]
This should probably be better-documented. Our output faceIDs are always indices into the input mesh's triangles vector. The idea is you have a vector of length input triangles that lists each face in a way you care about. Then you can put that vector in as the input faceID to get proper coplanarity, and you can also get your face by indexing into that vector using our output faceID. Does that work?

**Reply by ramcdona** [reply]
The problem I see is that there are usually more than one input mesh.  When you do a boolean operation, Mesh A and Mesh B are input, and Mesh C is output.

If I have a given triangle from Mesh C, and it has faceID=10, how do I know whether that triangle 'came from' Mesh A or Mesh B?

If C.faceID "is an index into the input mesh's triangle's vector", how do I know which one?

**Reply by elalish** [reply]
That's what `runOriginalID` is for.

**Reply by ramcdona** [reply]
Is there an example of runOriginalID's use anywhere?

From the documentation, I can't honestly tell whether it is an input or an output, or both.  It says to set ReserveIDs(), but I can't make heads or tails of what I'm supposed to set it to -- in particular for each mesh, for each triangle, or what?

I find it very counter intuitive that the user sets a faceID in -- and then that isn't returned as the same ID as output.  If the behavior isn't changed, I would suggest that it needs big warning labels around it in the documentation.

I still don't know what a 'run' is.

**Reply by elalish** [reply]
Okay, that's fair feedback - before we get into changing the API, let's get your project working. From the docs on runIndex: 

>Indicates runs of triangles that correspond to a particular input mesh instance. The runs encompass all of triVerts and are sorted by runOriginalID. Run i begins at triVerts[runIndex[i]] and ends at triVerts[runIndex[i+1]]. 

So a run is just a group of triangles that all came from a particular mesh input. The runOriginalID is both an input and an output, but is not a required input (you could just get the default-generated OriginalID from each Manifold you create). It's only useful to set multiple triangle runs as input if your mesh is already partitioned into pieces you care about, e.g. from earlier Boolean operations. The basic setup is to have each input mesh have only a single triangle run & ID, which is the default behavior if they aren't supplied. 

As for examples, I'd recommend our tests - [this](https://github.com/elalish/manifold/blob/master/test/test_main.cpp#L319) is the function used to verify that our output runs, IDs, transforms, properties, etc match those of the input meshes, so very much like what you're trying to achieve.

**Reply by ramcdona** [reply]
Thanks for all the help.

I've got something working based on this.





I think the easiest improvement would be to add some documentation (like what you've said here) defining what a run is and how it works.

Possibly include an example loop like this in the documentation.

`    manifold::Manifold mf;

    manifold::MeshGL m = mf.GetMeshGL();

    for ( int run = 0; run < m.runOriginalID.size(); ++run )

    {

        for ( int tri = m.runIndex[ run ] / 3; tri < m.runIndex[ run + 1 ] / 3; ++tri )

        {

            int inTri = m.faceID.empty() ? tri : m.faceID[ tri ];

            // inTri // index into input triangle array.

        }

    }`

Also add some documentation that MeshGL::faceID (as an output) is not related to the faceID's you input.  Instead, it is an index into the input triangle array.  From skimming the code, it looks like using faceID as an index into the input set is pretty entrenched and this wouldn't be easy to change.







If you decide to make more substantial changes, it feels like faceID's use as an input is very limited (to differentiate faces to prevent merger) vs. its use as an output.

I would propose splitting faceID into two variables (with likely renaming).

faceID (input) -- (rename to faceTag) this behaves the way I expected, it is just an integer marker (tag) that is carried through and any split triangles inherit their parent's faceTag.  This can serve the input face differentiation role that faceID input currently plays.

faceID (output) -- (rename to faceInputIndex) this behaves the way faceID currently does as an internal and output value.

For minimal changes, you would not rename faceID (output) to faceInputIndex -- you would leave it alone.  If someone wanted to use  this faceID as an input, it would need to be a simple array [0:nTri-1].

Just my $.02

**Reply by elalish** [reply]
Glad it's working! Agreed, the documentation definitely needs work and adding the sample loops is a good idea. It's true that faceID's use as an input is pretty limited - I need to look into that deeper. I want to ensure that you can save and reconstitute state with MeshGL losslessly, so the faceID thing may yet be a bug.

**Comment by ramcdona** [comment]
I'm digging deeper into the documentation, but it really doesn't make any sense...

I think the solution to my problem has something to do with:

MeshGL::runOriginalID[]
MeshGL::runIndex[]

Manifold::OriginalID()
Manifold::AsOriginal()
Manifold::ReserveIDs()

But these refer to different 'runs', meshid's, relations, MeshRelation -- which all seem to be terms that aren't defined anywhere I can find.

Is a 'run' a group of tris?  Or is it a pass through Manifold?

Are meshid's different from faceID's?  I don't see where meshes are assigned or return an ID.  Sentences like "The OriginalID of the mesh this triangle run came from." don't grok.

**Reply by elalish** [reply]
This is the relevant [documentation](https://manifoldcad.org/docs/html/structmanifold_1_1_mesh_g_l.html#aa67d01a9b240154a754945eca9fca3e0) - does this help? Happy to improve it where it's unclear.

**Reply by ramcdona** [reply]
I've read that 100 times, it makes no sense to me.

What is a 'run' of triangles?
